lstchain/reco/r0_to_dl1.py

```diff
@@ -535,9 +545,79 @@ def r0_to_dl1(
                         dl1_container.dragon_time = dragon_time_utc.unix
                         dl1_container.tib_time = tib_time_utc.unix
 
+                        # Until the TIB trigger_type is fully reliable, we also add
+                        # the ucts_trigger_type to the data
+                        dl1_container.ucts_trigger_type = event.lst.tel[telescope_id].evt.ucts_trigger_type
+
+                        # Due to a DAQ bug, sometimes there are 'jumps' in the
+                        # UCTS info in the raw files. After one such jump,
+                        # all the UCTS info attached to an event actually
+                        # corresponds to the next event. This one-event
+                        # shift stays like that until there is another jump
+                        # (then it becomes a 2-event shift and so on). We will
+                        # keep track of those jumps, by storing the UCTS info
+                        # of the previously read events in the list
+                        # previous_ucts_time_unix. The list has one element
+                        # for each of the jumps, so if there has been just
+                        # one jump we have the UCTS info of the previous
+                        # event only (which truly corresponds to the
+                        # current event). If there have been n jumps, we keep
+                        # the past n events. The info to be used for
+                        # the current event is always the first element of
+                        # the array, previous_ucts_time_unix[0], whereas the
+                        # current event's (wrong) ucts info is placed last in
+                        # the array. Each time the first array element is
+                        # used, it is removed and the rest move up in the
+                        # list. We have another similar array for the trigger
+                        # types, previous_ucts_trigger_type
+                        #
+                        if len(previous_ucts_time_unix) > 0:
+                            # keep the time & trigger type read for this
+                            # event (which really correspond to a later event):
+                            current_ucts_time = dl1_container.ucts_time
+                            current_ucts_trigger_type = dl1_container.ucts_trigger_type
+                            # put in dl1_container the proper time for this
+                            # event:
```

Comment:
This whole section is unnecessarily complex and could be improved a lot by making use of the methods python lists provide, i.e. pop and appe d. 

I think here, only a call `l.pop(0)` is needed.

It looks like collections.deque (double ended queue) would even be a better solution, as one only needs to popleft and append which is much more efficient and easier to read than these shifting loops.

Reply:
> This whole section is unnecessarily complex and could be improved a lot by making use of the methods python lists provide, i.e. pop and appe d.
> 
> I think here, only a call `l.pop(0)` is needed.
> 
> It looks like collections.deque (double ended queue) would even be a better solution, as one only needs to popleft and append which is much more efficient and easier to read than these shifting loops.

Well I guess it depends on one's command of python. It is explained, and not hard to read.

Reply:
You will be able to substitute this whole sections with just two small lines. And replace a very inefficient loop with idiomatic code that will run much faster:

This should do it (I did not test it though)
```
dl1_container.ucts_time = previous_ucts_times.pop(0)
previous_ucts_times.append(current_ucts_time)
```

Even more efficient would be, as I said, the use of a double ended queue:

```
from collections import deque

previous_ucts_times = deque()

...
dl1_container.ucts_time = previous_ucts_times.popleft()
previous_ucts_times.append(current_ucts_time)
```

Much cleaner, don't you agree?

Reply:
Thanks, I like the first version. Will give it a try.